Struct with generics fails on declaring []T

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 
OS: linux, Ubuntu 18.04.5 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-4200U CPU @ 1.60GHz
CC version: cc (Ubuntu 10.1.0-2ubuntu1~18.04) 10.1.0

getwd: /home/donp/projects/vbridge
vmodules: /home/donp/.vmodules
vroot: /opt/v/latest
vexe: /opt/v/latest/v
vexe mtime: 2021-03-30 19:43:44
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 dcab791

Git version: git version 2.31.0
Git vroot status: weekly.2021.11-214-gdcab7914
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81


```v
struct Queue<T> {
	queue []T
}

fn main() {
	q := Queue<int>{}
	println("queue elements $q.queue.len")
}
```


```v
error `/tmp/v/generics_struct.1922235238860865521.tmp.c:9217: error: 'T' undeclared`

VV_LOCAL_SYMBOL void main__main(void) {
9217:        main__Queue_T_int q = (main__Queue_T_int){.queue = __new_array(0, 1, sizeof(T)),};
        println(_STR("queue elements %"PRId32"", 1, q.queue.len));
}
```

--- a/vlib/v/table/table.v
+++ b/vlib/v/table/table.v
@@ -1091,27 +1091,22 @@ pub fn (mut t Table) generic_struct_insts_to_concrete() {
 			}
 			mut parent_info := parent.info as Struct
 			mut fields := parent_info.fields.clone()
-			for i, _ in fields {
-				mut field := fields[i]
-				if field.typ.has_flag(.generic) {
-					if parent_info.generic_types.len != info.generic_types.len {
-						// TODO: proper error
-						panic('generic template mismatch')
+			if parent_info.generic_types.len == info.generic_types.len {
+				for i, _ in fields {
+					mut field := fields[i]
+					if t_typ := t.resolve_generic_by_types(field.typ, parent_info.generic_types,
+						info.generic_types)
+					{
+						field.typ = t_typ
 					}
-					for j, gp in parent_info.generic_types {
-						if gp == field.typ {
-							field.typ = info.generic_types[j].derive(field.typ).clear_flag(.generic)
-							break
-						}
-					}
+					fields[i] = field
 				}
-				fields[i] = field
+				parent_info.generic_types = []
+				parent_info.fields = fields
+				typ.is_public = true
+				typ.kind = .struct_
+				typ.info = parent_info
 			}
-			parent_info.generic_types = []
-			parent_info.fields = fields
-			typ.is_public = true
-			typ.kind = .struct_
-			typ.info = parent_info
 		}
 	}
 }

--- a/vlib/v/tests/generics_return_generics_struct_test.v
+++ b/vlib/v/tests/generics_return_generics_struct_test.v
@@ -49,3 +49,15 @@ fn test_generics_with_generics_struct_string() {
 	assert ret.contains("data: ['foo', 'bar']")
 	assert ret.contains('index: 11')
 }
+
+fn test_generics_struct_insts_to_concrete() {
+	ai := ArrayIterator<int>{
+		data: [11, 22],
+		index: 22
+	}
+	println(ai)
+	ret := '$ai'
+	assert ret.contains('ArrayIterator<int>{')
+	assert ret.contains('data: [11, 22]')
+	assert ret.contains('index: 22')
+}

